backend/src/openai.ts

```diff
@@ -329,220 +369,127 @@ function getChatCompletionsFromHistory(
 }
 
-function pushCompletionToHistory(
-	chatHistory: ChatHistoryMessage[],
-	completion: ChatCompletionMessageParam,
-	chatMessageType: CHAT_MESSAGE_TYPE
-) {
-	// limit the length of the chat history
-	const maxChatHistoryLength = 1000;
-
-	if (chatMessageType !== CHAT_MESSAGE_TYPE.BOT_BLOCKED) {
-		// remove the oldest message, not including system role message
-		if (chatHistory.length >= maxChatHistoryLength) {
-			if (chatHistory[0].completion?.role !== 'system') {
-				chatHistory.shift();
-			} else {
-				chatHistory.splice(1, 1);
-			}
-		}
-		chatHistory.push({
-			completion,
-			chatMessageType,
-		});
-	} else {
-		// do not add the bots reply which was subsequently blocked
-		console.log('Skipping adding blocked message to chat history', completion);
-	}
-	return chatHistory;
-}
-
-function getBlankChatResponse(): ChatResponse {
-	return {
-		completion: null,
-		defenceReport: {
-			blockedReason: '',
-			isBlocked: false,
-			alertedDefences: [],
-			triggeredDefences: [],
-		},
-		wonLevel: false,
-		openAIErrorMessage: null,
-	};
-}
-
-function applyOutputFilterDefence(
-	message: string,
-	defences: Defence[],
-	chatResponse: ChatResponse
-) {
-	const detectedPhrases = detectFilterList(
-		message,
-		getFilterList(defences, DEFENCE_ID.FILTER_BOT_OUTPUT)
-	);
-
-	if (detectedPhrases.length > 0) {
-		console.debug(
-			'FILTER_BOT_OUTPUT defence triggered. Detected phrases from blocklist:',
-			detectedPhrases
-		);
-		if (isDefenceActive(DEFENCE_ID.FILTER_BOT_OUTPUT, defences)) {
-			chatResponse.defenceReport.triggeredDefences.push(
-				DEFENCE_ID.FILTER_BOT_OUTPUT
-			);
-			chatResponse.defenceReport.isBlocked = true;
-			chatResponse.defenceReport.blockedReason =
-				'Message Blocked: My response was blocked as it contained a restricted word/phrase.';
-		} else {
-			chatResponse.defenceReport.alertedDefences.push(
-				DEFENCE_ID.FILTER_BOT_OUTPUT
-			);
-		}
-	}
-}
-
 async function performToolCalls(
-	chatResponse: ChatResponse,
 	toolCalls: ChatCompletionMessageToolCall[],
 	chatHistory: ChatHistoryMessage[],
 	defences: Defence[],
-	sentEmails: EmailInfo[],
 	currentLevel: LEVEL_NAMES
-) {
+): Promise<ToolCallResponse> {
 	for (const toolCall of toolCalls) {
 		// only tool type supported by openai is function
 
 		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
 		if (toolCall.type === 'function') {
-			// call the function and get a new reply and defence info from
 			const functionCallReply = await chatGptCallFunction(
-				chatResponse.defenceReport,
 				defences,
 				toolCall.id,
 				toolCall.function,
-				sentEmails,
 				currentLevel
 			);
-			chatResponse.wonLevel = functionCallReply.wonLevel;
-
-			// add the function call to the chat history
-			pushCompletionToHistory(
-				chatHistory,
-				functionCallReply.completion,
-				CHAT_MESSAGE_TYPE.FUNCTION_CALL
-			);
-			// update the defence info
-			chatResponse.defenceReport = functionCallReply.defenceReport;
```

Comment:
It's all good!

Originally we were passing `defenceReport` into `chatGptCallFunction`, and doing precisely zero processing on it / with it before spitting it out again in the return object. That is to say, we never needed to pass it in or out in the first place. After all, the only things we'd do in `chatGptCallFunction` is look at docs or send an email, and currently there are no defences that interact with those systems.

Now, all the defence detection calls live inside `handleChatWithDefenceDetection` in _chatController.ts_. Which is nice because `chatGptSendMessage` is now solely responsible for getting a reply from gpt and dealing with function calls. The input defence detection happens before that, and the output defence detection happens after, and all the concerns are happily separated (mostly).

Reply:
Thank you, excellent explanation! You can tell I wrote this comment on a friday afternoon of a hectic week, cos I missed that all-important detail that we weren't even doing anything with it in the original call 😅